Library/include/CSP/Multiplayer/EventBus.h

```diff
@@ -0,0 +1,110 @@
+	friend class csp::systems::SystemsManager;
+	friend class MultiplayerConnection;
+	friend void csp::memory::Delete<EventBus>(EventBus* Ptr);
+	/** @endcond */
+
+	// The callback used to register to listen to events.
+	typedef std::function<void(bool, const csp::common::Array<ReplicatedValue>&)> ParameterisedCallbackHandler;
+
+	/// @brief Registers a system to listen for the named event, where the system can define its
+	/// @brief own callback and deserialiser.
+	/// @param EventName csp::common::String : The identifying name for the event to listen for.
+	/// @param System csp::systems::SystemBase* : A pointer to the system which wants to register for the event.
+	CSP_NO_EXPORT void ListenEvent(const csp::common::String& EventName, csp::systems::SystemBase* System);
+
+	/// @brief Registers a callback to listen for the named event
+	/// @param EventName csp::common::String : The identifying name for the event to listen for.
+	/// @param Callback ParameterisedCallbackHandler : A callback to register for the event which contains the parameter payload data.
+	void ListenEvent(const csp::common::String& EventName, ParameterisedCallbackHandler Callback);
+
+	/// @brief Stops the event bus from listening for a particular event, for any system or callbacks
+	/// @brief that were registered.
+	/// @param EventName csp::common::String : The identifying name for the event to stop listening for.
+	void StopAllListenEvent(const csp::common::String& EventName);
+
+	/// @brief Stops the event bus from listening for a particular event, for the specified system.
+	/// @param EventName csp::common::String : The identifying name for the event to stop listening for.
+	/// @param System csp::systems::SystemBase : The system that will stop being registered.
+	void StopSystemListenEvent(const csp::common::String& EventName, csp::systems::SystemBase* System);
+
+	/// @brief Stops the event bus from listening for a particular event, for the specified callback.
+	/// @brief Other callbacks which may be registered to the same event are not affected.
+	/// @param EventName csp::common::String : The identifying name for the event to stop listening for.
+	/// @param Callback ParameterisedCallbackHandler : The callback that will stop being registered.
+	void StopCallbackListenEvent(const csp::common::String& EventName, ParameterisedCallbackHandler Callback);
+
+	/// @brief Instructs the event bus to start listening to messages
+	void StartEventMessageListening();
+
+private:
+	EventBus();
+	~EventBus();
+
+	EventBus(MultiplayerConnection* InMultiplayerConnection);
+
+	class MultiplayerConnection* MultiplayerConnectionInst;
+
+	// TODO: Replace these with pointers! We can't use STL containers as class fields due to the fact that the class size will
+	//   change depending on which runtime is used.
+	typedef std::vector<ParameterisedCallbackHandler> Callbacks;
+	typedef std::map<csp::common::String, std::pair<Callbacks, csp::systems::SystemBase*>> EventMaps;
```

Comment:
I think separating would be more clear and easier to handle

Reply:
the rationale for the single container was to support registering both system(s) and callback(s) for the same event, but since then we've agreed that this was off the table. I agree that in the case where we only support either/or, 2 containers is probably best

Reply:
Yes ok makes sense. Thanks

Reply:
Yeah it that sounds good, and think it would help improve readability and clarity